pkg/cloud/aws/services/userdata/controlplane.go

```diff
@@ -16,111 +16,187 @@ limitations under the License.
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.FrontProxyCAKey | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/sa.pub
+    encoding: "base64"
+    owner: root:root
+    permissions: '0640'
+    content: |
+      {{.SaCert | Base64Encode}}
+
+-   path: /etc/kubernetes/pki/sa.key
+    encoding: "base64"
+    owner: root:root
+    permissions: '0600'
+    content: |
+      {{.SaKey | Base64Encode}}
+
+-   path: /tmp/kubeadm.yaml
+    owner: root:root
+    permissions: '0640'
+    content: |
+      ---
+      apiVersion: kubeadm.k8s.io/v1beta1
+      kind: ClusterConfiguration
+      apiServer:
+        certSANs:
+          - {{ "{{ ds.meta_data.local_ipv4 }}" }}
+          - "{{.ELBAddress}}"
+        extraArgs:
+          cloud-provider: aws
+      controlPlaneEndpoint: "{{.ELBAddress}}:6443"
+      clusterName: "{{.ClusterName}}"
+      networking:
+        dnsDomain: "{{.ServiceDomain}}"
+        podSubnet: "{{.PodSubnet}}"
+        serviceSubnet: "{{.ServiceSubnet}}"
+      kubernetesVersion: "{{.KubernetesVersion}}"
+      ---
+      apiVersion: kubeadm.k8s.io/v1beta1
+      kind: InitConfiguration
+      nodeRegistration:
+        name: {{ "{{ ds.meta_data.hostname }}" }}
+        criSocket: /var/run/containerd/containerd.sock
+        kubeletExtraArgs:
+          cloud-provider: aws
+kubeadm:
+  operation: init
+  config: /tmp/kubeadm.yaml
+
 `
 
-	controlPlaneJoinBashScript = `{{.Header}}
-
-set -eox
-
-mkdir -p /etc/kubernetes/pki/etcd
-
-echo -n '{{.CACert}}' > /etc/kubernetes/pki/ca.crt
-echo -n '{{.CAKey}}' > /etc/kubernetes/pki/ca.key
-chmod 600 /etc/kubernetes/pki/ca.key
-
-echo -n '{{.EtcdCACert}}' > /etc/kubernetes/pki/etcd/ca.crt
-echo -n '{{.EtcdCAKey}}' > /etc/kubernetes/pki/etcd/ca.key
-chmod 600 /etc/kubernetes/pki/etcd/ca.key
-
-echo -n '{{.FrontProxyCACert}}' > /etc/kubernetes/pki/front-proxy-ca.crt
-echo -n '{{.FrontProxyCAKey}}' > /etc/kubernetes/pki/front-proxy-ca.key
-chmod 600 /etc/kubernetes/pki/front-proxy-ca.key
-
-echo -n '{{.SaCert}}' > /etc/kubernetes/pki/sa.pub
-echo -n '{{.SaKey}}' > /etc/kubernetes/pki/sa.key
-chmod 600 /etc/kubernetes/pki/sa.key
-
-PRIVATE_IP=$(curl http://169.254.169.254/latest/meta-data/local-ipv4)
-HOSTNAME="$(curl http://169.254.169.254/latest/meta-data/local-hostname)"
-
-cat >/tmp/kubeadm-controlplane-join-config.yaml <<EOF
----
-apiVersion: kubeadm.k8s.io/v1beta1
-kind: JoinConfiguration
-discovery:
-  bootstrapToken:
-    token: "{{.BootstrapToken}}"
-    apiServerEndpoint: "{{.ELBAddress}}:6443"
-    caCertHashes:
-      - "{{.CACertHash}}"
-nodeRegistration:
-  name: "${HOSTNAME}"
-  criSocket: /var/run/containerd/containerd.sock
-  kubeletExtraArgs:
-    cloud-provider: aws
-controlPlane:
-  localAPIEndpoint:
-    advertiseAddress: "${PRIVATE_IP}"
-    bindPort: 6443
-EOF
-
-kubeadm join --config /tmp/kubeadm-controlplane-join-config.yaml --v 10
+	controlPlaneJoinCloudInit = `{{.Header}}
```

Comment:
Have you tested joining an additional control plane instance? I don't believe this will work as is, because the `--experimental-control-plane` is needed when invoking `kubeadm`.

My initial thoughts are that we can update the cloud-init module to add a separate operation for control plane join, which could then pass the required flag. It might need some special handling (for v1.14+, since the flag will no longer be experimental, though.

Reply:
It seems to work. I tested it by adding additional plane node (controlplane-1) to machines.yaml file and running apply-machines phase. It joins fine.

Reply:
I'll take it :man_shrugging:. I'll attempt to give this (and #538) a test this afternoon.

Reply:
From the diff it looks like the bash version was just joining the plane without any special options.

Reply:
Yeah, the `--experimental-control-plane` is required when the join params are passed on the cli. We are using the `JoinConfiguration` to invoke `kubeadm join`.